query-algebrizer/src/clauses/hints.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+
+use clauses::{
+    ConjoiningClauses,
+};
+
+use mentat_query::{
+    FnArg,
+    Predicate,
+};
+
+use mentat_core::{
+    ValueType,
+    ValueTypeSet,
+};
+
+use errors::{
+    ErrorKind,
+    Result,
+};
+
+use types::{
+    EmptyBecause,
+};
+
+impl ConjoiningClauses {
+
+    pub fn apply_type_hint(&mut self, pred: Predicate, ty: ValueType) -> Result<()> {
+        if pred.args.len() != 1 {
+            bail!(ErrorKind::InvalidNumberOfArguments(pred.operator.clone(), pred.args.len(), 1));
+        }
+
+        if let FnArg::Variable(ref v) = pred.args[0] {
+            if let Some(existing) = self.type_hints.insert(v.clone(), ty) {
```

Comment:
I agree with your naming concern, and I think I would call this `required_types` instead of `type_hints`.

Reply:
Refactoring request: make this a separate method that takes a type and a variable. Leave the arg parsing to this function. We will use this functionality elsewhere (e.g., to fix #520).

Reply:
Change request: make `type_hints` (aka `required_types`) a `ValueTypeSet`, not a `ValueType`. That way we can impose both numeric types.